File: labs/control-ui/api-shadow/dynamic_proxy_validator.ts
// Control UI TypeScript: Dynamic Proxy Pool Validator API

export type ProxyProtocol = 'http' | 'https' | 'socks4' | 'socks5';
export type AnonymityLevel = 'transparent' | 'anonymous' | 'elite';

export interface ProxyRecord {
  host: string;
  port: number;
  protocol: ProxyProtocol;
  isAlive: boolean;
  latencyMs: number;
  anonymity: AnonymityLevel;
  lastCheckedMs: number;
}

export class DynamicProxyValidator {
  public timeoutMs: number;
  public pool: Map<string, ProxyRecord> = new Map();

  constructor(timeoutMs: number = 5000) {
    this.timeoutMs = timeoutMs;
  }

  public craftSocks5Probe(): Uint8Array {
    return new Uint8Array([0x05, 0x01, 0x00]);
  }

  public craftHttpProbe(host: string): string {
    return `GET http://${host}/generate_204 HTTP/1.1\r\nHost: ${host}\r\nConnection: close\r\n\r\n`;
  }

  public verifySocks5Response(resp: Uint8Array): boolean {
    return resp.length >= 2 && resp[0] === 0x05 && resp[1] === 0x00;
  }

  public determineAnonymity(headers: Record<string, string>, clientIp: string): AnonymityLevel {
    const keys = ['x-forwarded-for', 'via', 'x-real-ip', 'forwarded'];
    let hasForward = false;

    for (const [k, v] of Object.entries(headers)) {
      if (keys.includes(k.toLowerCase())) {
        hasForward = true;
        if (v.includes(clientIp)) return 'transparent';
      }
    }

    return hasForward ? 'anonymous' : 'elite';
  }

  public recordProbe(
    host: string,
    port: number,
    protocol: ProxyProtocol,
    isAlive: boolean,
    latencyMs: number,
    anonymity: AnonymityLevel,
    nowMs: number
  ): void {
    const key = `${host}:${port}`;
    this.pool.set(key, {
      host,
      port,
      protocol,
      isAlive,
      latencyMs,
      anonymity,
      lastCheckedMs: nowMs,
    });
  }

  public getHealthyProxies(maxLatencyMs: number): ProxyRecord[] {
    return Array.from(this.pool.values())
      .filter(p => p.isAlive && p.latencyMs <= maxLatencyMs)
      .sort((a, b) => a.latencyMs - b.latencyMs);
  }
}

File: labs/mobile/donor-archive/tunnel/DynamicProxyValidator.kt
package com.luminet.android.tunnel

import java.util.concurrent.ConcurrentHashMap

enum class AndroidProxyProto {
    HTTP,
    HTTPS,
    SOCKS4,
    SOCKS5
}

enum class AndroidProxyAnon {
    TRANSPARENT,
    ANONYMOUS,
    ELITE
}

data class AndroidProxyRecord(
    val host: String,
    val port: Int,
    val protocol: AndroidProxyProto,
    val isAlive: Boolean,
    val latencyMs: Long,
    val anonymity: AndroidProxyAnon,
    val lastCheckedMs: Long
)

class DynamicProxyValidator(val timeoutMs: Long = 5000) {
    private val pool = ConcurrentHashMap<String, AndroidProxyRecord>()

    fun craftSocks5Probe(): ByteArray = byteArrayOf(0x05, 0x01, 0x00)

    fun craftHttpProbe(host: String): String =
        "GET http://$host/generate_204 HTTP/1.1\r\nHost: $host\r\nConnection: close\r\n\r\n"

    fun verifySocks5Response(resp: ByteArray): Boolean =
        resp.size >= 2 && resp[0] == 0x05.toByte() && resp[1] == 0x00.toByte()

    fun determineAnonymity(headers: Map<String, String>, myIp: String): AndroidProxyAnon {
        val forwardKeys = listOf("x-forwarded-for", "via", "x-real-ip", "forwarded")
        var hasForwarding = false

        for ((k, v) in headers) {
            if (forwardKeys.contains(k.lowercase())) {
                hasForwarding = true
                if (v.contains(myIp)) return AndroidProxyAnon.TRANSPARENT
            }
        }

        return if (hasForwarding) AndroidProxyAnon.ANONYMOUS else AndroidProxyAnon.ELITE
    }

    fun recordProbe(
        host: String,
        port: Int,
        proto: AndroidProxyProto,
        isAlive: Boolean,
        latencyMs: Long,
        anon: AndroidProxyAnon,
        nowMs: Long
    ) {
        val key = "$host:$port"
        pool[key] = AndroidProxyRecord(host, port, proto, isAlive, latencyMs, anon, nowMs)
    }

    fun getHealthyProxies(maxLatencyMs: Long): List<AndroidProxyRecord> {
        return pool.values
            .filter { it.isAlive && it.latencyMs <= maxLatencyMs }
            .sortedBy { it.latencyMs }
    }
}
